capsules/src/process_console.rs

```diff
@@ -1033,37 +1090,29 @@ impl<'a, C: ProcessManagementCapability> uart::TransmitClient for ProcessConsole
         _tx_len: usize,
         _rcode: Result<(), ErrorCode>,
     ) {
+        // Reset state now that we no longer have an active transmission on the
+        // UART.
         self.tx_buffer.replace(buffer);
         self.tx_in_progress.set(false);
 
-        // If in the middle of an active state, finish the state machine.
-        if self.writer_state.get() != WriterState::Empty
-            && self.writer_state.get() != WriterState::KernelStart
-            && self.writer_state.get() != WriterState::ProcessStart
-        {
-            self.write_state(WriterState::Empty, None);
-        }
+        // Check if we have anything queued up. If we do, let the queue
+        // empty.
+        let ret = self.handle_queue();
+        if ret.ok() == Some(0) || ret.is_err() {
```

Comment:
It seems like this logic is such that if the underlying UART is busy during this callback (https://github.com/tock/tock/blob/f33ce870aea456acaf40c2a66ec880d2142e1249/capsules/src/process_console.rs#L1046), then this will lead this to enqueue more. I'm wondering if the `is_err()` case is correct here. Can you walk me through how this is supposed to work?

Reply:
The intent is: If for whatever reason handling the queue did not cause a UART TX (and therefore we will not get another `transmitted_buffer()` callback), then we continue handling other work that might need to use the UART TX.

Since we set `tx_in_progress` in progress to false immediately before this, handle_queue can't fail due to the TX being busy in this case. Perhaps there could be other failure cases in handle_queue in the future.